### PR TITLE
syntheticprivilegecache: use correct context when reading the table

### LINKED ISSUE #126979: roachtest: multitenant failed: use of Span after Finish
roachtest.multitenant/distsql/instances=20/bundle=on/timeout=1 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/16022118?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/16022118?buildTab=artifacts#/multitenant/distsql/instances=20/bundle=on/timeout=1) on release-24.2 @ [495d20e82f09b3436f1b00d7331146fe2042dd6a](https://github.com/cockroachdb/cockroach/commits/495d20e82f09b3436f1b00d7331146fe2042dd6a):

```
(assertions.go:363).Fail: 
	Error Trace:	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/multitenant_distsql.go:193
	            				github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/multitenant_distsql.go:47
	            				main/pkg/cmd/roachtest/test_runner.go:1248
	            				src/runtime/asm_amd64.s:1695
	Error:      	Error message not equal:
	            	expected: "pq: query execution canceled due to statement timeout"
	            	actual  : "dial tcp 104.196.204.199:29000: connect: connection refused"
	Test:       	multitenant/distsql/instances=20/bundle=on/timeout=1
(require.go:177).EqualError: FailNow called
test artifacts and logs in: /artifacts/multitenant/distsql/instances=20/bundle=on/timeout=1/run_1
```

Parameters:
 - <code>ROACHTEST_arch=amd64</code>
 - <code>ROACHTEST_cloud=gce</code>
 - <code>ROACHTEST_coverageBuild=false</code>
 - <code>ROACHTEST_cpu=4</code>
 - <code>ROACHTEST_encrypted=false</code>
 - <code>ROACHTEST_metamorphicBuild=false</code>
 - <code>ROACHTEST_ssd=0</code>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-16022118/multitenant-distsql-instances-20-bundle-on-timeout-1/1720678849138/1720679294158)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #126313 roachtest: multitenant/distsql/instances=20/bundle=on/timeout=1 failed [C-test-failure O-roachtest O-robot T-sql-queries branch-master release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*multitenant/distsql/instances=20/bundle=on/timeout=1.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

Jira issue: CRDB-40216

### PR DESCRIPTION
**Pull request by yuzefovich**
Previously, we incorrectly used the captured context in `readFromStorage` invocation which happens in a separate goroutine which might outlive its creator. This in turn could lead to a "span use after Finish" problem and is now fixed.

Fixes: #126979.

Release note: None

### CODE REVIEW

**Review by mgartner — APPROVED** [review]
Nice find! How sure are you that this is causing #126979? I'm wondering why we've started to see this error only now, since this code has been like this for years.

**Comment by yuzefovich** [comment]
I'm not too confident that this is the fix, but I'm quite optimistic about it. Looks like the first occurrence of the problem is the first message in #126313 which happened soon after #125590 merged. I'm wondering whether #125590 somehow made this bug to be more likely.

I'll definitely hold off on merging the backports if we want to do that, but let's see what happens in the nightly on master.

TFTR!

bors r+

**Pull request merged**
